pkg/expr/graph.go

```diff
@@ -55,7 +55,45 @@ type DataPipeline []Node
 // map of the refId of the of each command
 func (dp *DataPipeline) execute(c context.Context, now time.Time, s *Service) (mathexp.Vars, error) {
 	vars := make(mathexp.Vars)
+
+	// Execute datasource nodes first, and grouped by datasource.
+	dsNodes := []*DSNode{}
+	for _, node := range *dp {
+		if node.NodeType() != TypeDatasourceNode {
+			continue
+		}
+		dsNodes = append(dsNodes, node.(*DSNode))
+	}
+
+	executeDSNodesGrouped(c, now, vars, s, dsNodes)
+
+	// TODO: Mark Dependent nodes of vars[X] that have result.Error != nil, and then don't execute dependent nodes,
+	// but instead set them to an error and return
+
 	for _, node := range *dp {
+		if node.NodeType() == TypeDatasourceNode {
+			continue // already executed via executeDSNodesGrouped
+		}
+
+		var hasDepError bool
+		if cmd, ok := node.(*CMDNode); ok {
+			for _, neededVar := range cmd.Command.NeedsVars() {
+				if res, ok := vars[neededVar]; ok {
+					if res.Error != nil {
+						errResult := mathexp.Results{
+							Error: fmt.Errorf("not executing expression %v because expression %v errored", node.RefID(), neededVar), // TODO errutil
```

Comment:
It can be a public variable so in alerting we are able to filter them out and report only the root cause.

Reply:
I adjusted eval.go in alerting to walk the Dependency Errors and return the underlying non-dependency error.